resolver/src/lookup_state.rs

```diff
@@ -239,61 +240,54 @@ impl<C: DnsHandle<Error = ResolveError> + 'static> QueryFuture<C> {
-                    search_name = name.clone();
-                }
-            } else {
-                // now that is very odd...
-                warn!("Expected RData::CNAME in response record {:?}", cname);
-                break;
+        let (search_name, cname_ttl, was_cname) = {
+            let (search_name, cname_ttl, was_cname) = message.answers().iter()
+                .fold((Cow::Borrowed(self.query.name()), 0, false),
+                      |(search_name, cname_ttl, was_cname), r| {
+                          if let &RData::CNAME(ref cname) = r.rdata() {
+                              debug_assert_eq!(r.rr_type(), RecordType::CNAME);
+                              if search_name.as_ref() == r.name() {
+                                  return (Cow::Owned(cname.clone()), r.ttl(), true);
+                              }
+                          }
+                          (search_name, cname_ttl, was_cname)
+                      });
+
+            // After following all the CNAMES to the last one, try and lookup the final name
+            let records = message
+                .take_answers()
+                .into_iter()
+                .chain(message.take_additionals().into_iter())
+                .filter_map(|r| {
+                    let ttl = r.ttl();
+                    // TODO: disable name validation with ResolverOpts?
+                    // restrict to the RData type requested
+                    if self.query.query_type() == r.rr_type() && search_name.as_ref() == r.name() {
+                        Some((r.unwrap_rdata(), ttl))
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            if !records.is_empty() {
+                return Ok(Async::Ready(Records::Exists(records)));
             }
-        }
 
-        // After following all the CNAMES to the last one, try and lookup the final name
-        let records = message
-            .take_answers()
-            .into_iter()
-            .chain(message.take_additionals().into_iter())
-            .filter_map(|r| {
-                let ttl = r.ttl();
-                // TODO: disable name validation with ResolverOpts?
-                // restrict to the RData type requested
-                if self.query.query_type() == r.rr_type() && &search_name == r.name() {
-                    Some((r.unwrap_rdata(), ttl))
-                } else {
-                    None
-                }
-            })
-            .collect::<Vec<_>>();
+            (search_name.into_owned(), cname_ttl, was_cname)
+        };
 
-        if !records.is_empty() {
-            Ok(Async::Ready(Records::Exists(records)))
+        // It was a CNAME, but not included in the request...
+        if was_cname {
+            let next_query = Query::query(search_name, self.query.query_type());
+            Ok(Async::Ready(
+                self.next_query(next_query, cname_ttl, message),
```

Comment:
Why do we pass `cname_ttl` here? In particular, in a response that has three CNAME entries chained, why do we only care about the last CNAME entry's TTL? Like I mentioned above, it seems like we need to update the cache with all the CNAMEs in the chain so far, each with their own TTL, before we recurse with `next_query()`. But maybe I'm misunderstanding something.

Reply:
This is a great point about TTL. See my comment above about actually caching CNAMEs in the chain. As to TTL, this is probably wrong. We can discuss this, but given that it's a chain of CNAMEs, we should probably take the minimum TTL of all the CNAMEs in the chain as the actual TTL to store with the cached item. I think when I wrote this I only considered the original CNAME TTL as the TTL for the record...

By the way TTL during a query should always be > 0. I treat try to treat TTL more generally as a hint, that is, in some cases if it's out of bounds from what we decide it should be with regards to minimums, etc., it can be ignored. I believe this is also the recommendation of the RFCs, if I remember correctly. Does that make sense?